[Excel] (Custom functions) How to add a description using autogenerated metadata (#1055)
OfficeDev/office-js-docs-pr: docs/excel/custom-functions-json-autogeneration.md

---
ms.date: 06/03/2019

> description: Use JSDOC tags to dynamically create your custom functions JSON metadata.

Author: ```suggestion
description: Use JSDoc tags to dynamically create your custom functions JSON metadata.
```

#### id

> The `id` is used as the invariant identifier for the custom function stored in the document. It should not change.

Reviewer: ```suggestion
The `id` is an invariant identifier for the custom function.
```
Not sure what "stored in the document" means. What document?
 "it should not change" is redundant.

* If `id` is not provided, the JavaScript/TypeScript function name is converted to uppercase and disallowed characters are removed.
* The `id` must be unique for all custom functions.
* The allowed characters are limited to: A-Z, a-z, 0-9, underscores (\_), and period (.).

#### name

Provides the display `name` for the custom function.

* If name is not provided, the id is also used as the name.
* Allowed characters: Letters [Unicode Alphabetic character](https://www.unicode.org/reports/tr44/tr44-22.html#Alphabetic), numbers, period (.), and underscore (\_).
* Must start with a letter.
* Maximum length is 128 characters.

> ### description

Reviewer: This heading looks a bit odd interspersed between the other tags (as though it should also have a tag). I'd move this up to be the first discussion item before getting into tags. Two reasons.
1. It won't be in between the other tags.
2. JSDoc almost always starts with the description. And we've been doing the same with our samples in Script lab. I think we should consistently do this with all samples.

> A description does not require any specific tag. Add a description to a custom function by adding a phrase to describe what the function does inside the JSDoc comment. By default, whatever text is untagged in the JSDoc comment section will be the description of the function. The description appears to users in Excel as they are entering the function. In the following example, the phrase "A function that sums two numbers" is the description for the custom function with the id property of `SUM`.

Reviewer: ```suggestion
The description is displayed to the user as help text when they need help to understand what your custom function does. The description does not require any specific tag. Just enter a short text description in the JSDoc comment. In general the description is placed at the start of the JSDoc comment section. But it will work no matter where it is placed.

To see examples of the built-in function descriptions, open Excel, go to the **Formulas** tab, and choose **Insert function**. You can then browse through all the function descriptions, and also see your own custom functions listed.

In the following example, the phrase "Calculates the volume of a sphere." is the description for the custom function.
```

/**
 * Calculates the volume of a sphere.
 * @customfunction
...
*/

Don't want to use SUM as that is a built-in function with a different description which would be confusing.
